src/coreclr/jit/fgopt.cpp

```diff
@@ -4522,6 +4537,314 @@ bool Compiler::fgReorderBlocks(bool useProfile)
+            //
+            if ((enclosingTryEnd == nullptr) || enclosingTryEnd->NextIs(HBtab->ebdTryBeg))
+            {
+                tryRegionEnds[enclosingTryIndex] = tryEnd;
+            }
+        }
+    }
+
+    // Now, do the handler regions
+    //
+    XTnum = 0;
+    for (EHblkDsc* const HBtab : EHClauses(this))
+    {
+        // The end of each handler region should have been visited by iterating the blocklist above
+        //
+        BasicBlock* const hndEnd = hndRegionEnds[XTnum++];
+        assert(hndEnd != nullptr);
+
+        // Update the end pointer of this handler region to the new last block
+        //
+        HBtab->ebdHndLast                = hndEnd;
+        const unsigned enclosingHndIndex = HBtab->ebdEnclosingHndIndex;
+
+        // If this handler region is nested in another one, we might need to update its enclosing region's end block
+        //
+        if (enclosingHndIndex != EHblkDsc::NO_ENCLOSING_INDEX)
+        {
+            BasicBlock* const enclosingHndEnd = hndRegionEnds[enclosingHndIndex];
+            assert(enclosingHndEnd != nullptr);
+
+            // If the enclosing region ends right before the nested region begins,
+            // extend the enclosing region's last block to the end of the nested region.
+            //
+            BasicBlock* const hndBeg = HBtab->HasFilter() ? HBtab->ebdFilter : HBtab->ebdHndBeg;
+            if (enclosingHndEnd->NextIs(hndBeg))
+            {
+                hndRegionEnds[enclosingHndIndex] = hndEnd;
+            }
+        }
+    }
+}
+
+//-----------------------------------------------------------------------------
+// fgMoveColdBlocks: Move rarely-run blocks to the end of their respective regions.
+//
+// Notes:
+//    Exception handlers are assumed to be cold, so we won't move blocks within them.
+//
+void Compiler::fgMoveColdBlocks()
+{
+#ifdef DEBUG
+    if (verbose)
+    {
+        printf("*************** In fgMoveColdBlocks()\n");
+
+        printf("\nInitial BasicBlocks");
+        fgDispBasicBlocks(verboseTrees);
+        printf("\n");
+    }
+#endif // DEBUG
+
+    // In a single pass, find the end blocks of each region that we can use as insertion points for cold blocks
+    //
+    BasicBlock** const lastColdTryBlocks =
+        (compHndBBtabCount == 0) ? nullptr : new (this, CMK_Generic) BasicBlock* [compHndBBtabCount] {};
+    BasicBlock* const lastMainBB        = fgLastBBInMainFunction();
+    BasicBlock*       lastColdMainBlock = nullptr;
+
+    for (BasicBlock* const block : Blocks(fgFirstBB, lastMainBB))
+    {
+        // If a region ends with a call-finally pair, don't split the pair up.
+        // Instead, set the end of the region to the BBJ_CALLFINALLY block in the pair.
+        // Also, don't consider blocks in handler regions.
+        // (If all of some try region X is enclosed in an exception handler,
+        // lastColdTryBlocks[X] will be null. We will handle this case later.)
```

Comment:
Maybe be a bit more specific? 
```suggestion
        // lastColdTryBlocks[X] will be null. We will check for this case in tryMovingBlocks below.)
```